submariner/cable-driver-policy.md

```diff
@@ -0,0 +1,165 @@
+# Cable Driver Policy Support
+
+## Summary
+
+Submariner currently offers three cable drivers: VXLAN, IPSec or wireguard. However only one of these cable drivers
+is enabled across all the clusters in a ClusterSet; That is all clusters that join a broker must use the same cable
+driver. The goal of this proposal is to extend the current model to support the selection of different cable types
+between connected clusters.
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+ClusterSet. For example, a ClusterSet could contain three clusters, two that are on-premise and the other
+in a public cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+
+## Proposal
+
+This enhancement proposes a new policy to allow users to selectively deploy different cable types between clusters
+in a ClusterSet. The proposal assumes that connection between two clusters is symmetrical. If Cluster A is connected
+to Cluster B via a VXLAN cable type - then Cluster B is connected to Cluster A via a VXLAN cable type and not another
+cable type.
+
+This enhancement will propose a single API for cable driver policy support using custom resource definition (CRDs).
+
+### Cable driver policy API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    ClusterConnectionPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectionPoliciesSpec struct {
+ 
+    Policies []ClusterConnectionPolicy `json:"policies,omitempty"`
+}
+
+type ClusterConnectionPolicy struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // Name of the cable driver implementation to use for this connection.
+    CableDriverName string `json:"CableDriverName"`
+
+    // CableOptions specifies a additional configuration parameters for the cable driver.
+    CableOptions map[string]string `json:"CableOptions,omitempty"`
```

Comment:
@maryamtahhan I was wondering how you envision this being used. This is included in the policy spec so that would imply that we could have 2 or more policies referencing the same driver name but different options, eg

```yaml
  leftClusterSelector:
    env: production
  rightClusterSelector:
    env: production
  cableDriverName:
    name:  "ipsec"
  cableDriverOptions:
    foo: 123

  leftClusterSelector:
    env: testing
  rightClusterSelector:
    env: testing
  cableDriverName:
    name:  "ipsec"
  cableDriverOptions:
    foo: 456
```

Would we then spin up 2 instances of the driver, each with its own config? Is it possible to have 2 instances of the same driver running in a GW? 

Or should there only be one set of config options for a driver? If so then it seems the options shouldn't be defined here but in a separate `ConfigMap`.

Reply:
Perhaps we should just remove the cable driver configuration from this proposal altogether and leave it as a possible future enhancement.